### DISCUSSION TITLE
Suspense boundary is not working allowing for runtime error

### DISCUSSION THREAD

**Discussion by RIP21** [urql-graphql/urql#3243] [Q&A] [answered]
In one of the spots of my code, I have a query that just fetches the list of the items. And for some reason, it allows for components to run their code resulting in runtime errors before the stuff gets fetched, while they are supposed to be suspended before the results are returned.

In order for Typescript to be less strict and always assume `data` is not `undefined` I use such a helper hook.

```ts
export const useQueryData = <
  Data = unknown,
  Variables extends AnyVariables = AnyVariables,
>(
  args: UseQueryArgs<Variables, Data>,
) => {
  if (process.env.NODE_ENV === "development") {
    if (args.pause) {
      throw Error(
        "You should not use useQueryData hook in the components that PAUSE the query as it will lead to runtime error. Use useQuery instead.",
      )
    }
  }
  const [opResult, reexecute] = useUrqlQuery<Data, Variables>(args)
  // We throw instantly for ErrorBoundary to handle that. No handling in the component itself
  if (opResult.error) {
    throw opResult.error
  }
  return [opResult.data, reexecute, opResult] as [
    Exclude<typeof opResult.data, undefined>,
    typeof reexecute,
    typeof opResult,
  ]
}
```

So then I use it like
```ts

const [data] =  useQueryData({ query: UseItem_ItemsDocument })
const items = data.items
//                ^ here I have a runtime error
```

Despite the component that uses that being wrapped in Suspense it fails.
In development, it NEVER happens, but in a production build, it happens all the time.

I use `next-urql` and the latest versions of `urql` and `graphcache` available. This happens client-side, on server-side it works just fine.

Any ideas what can be wrong?

**Comment by RIP21** [comment]
Apologies, probably it's irrelevant. Do not act just yet :) Double checking something :)

**Reply by RIP21** [reply, marked as answer]
Yeah, irrelevant. Sorry :)

**Reply by kitten** [reply]
No worries!
Take care though that `data` can still be `null` when Graphcache encounters an unresolveable error. This should of course optimally only be limited to development, but just to let you know, we don't yet add errors for Graphcache 😅

**Reply by RIP21** [reply]
@kitten unresolvable error being?
Like GQL or Network error is expected, I guess, so my helper hook above will do the trick, but still.

BTW, what is the behaviour when the cache is getting invalidated (using updates of Graphcache)? All active operations are going to refetch and while they do old data will be shown, right?
The same applies to client recreation, what's the behaviour? Is it safe to assume `data` still be there, or, at least, suspend, while it's all in progress?

I partially invalidate part of the cache quite often using some dummy mutations (a hack for now), but very rarely, I still get `Cannot read properties of null (reading 'redacted') error` which is quite annoying as I expect Suspense to be reliable and code to be less defensive yet I still encounter `null`s and `undefined` sometimes.
